### PR TITLE
Suppress slab_normal warning.

### PR DESCRIPTION
**Pull request by LZaw**
Fixes second half of #2507.
Implemented the `always_poison` for vectors and used them on the slab_normal to suppress the compiler warning.
`always_poison` as well as `PoisonImpl<T>::do_always_posion` are only implemented for
- `float`
- `double`
- `Vector`

All the other classes are still open to implement this, as soon as they need it.

This Pullrequest may still contain changes made by #2508.

### CODE REVIEW

**Comment by oktomus** [comment]
The appveyor job completely failed. We should restart it. @FrancoisBeaune ? :)

File: src/appleseed/foundation/utility/poison.h [hunk 1]
```diff
@@ -112,6 +121,10 @@ class PoisonImpl<float>
     {
         x = FP<float>::snan();
     }
+    static void do_always_poison(float& x)
```
Comment (oktomus) [inline src/appleseed/foundation/utility/poison.h]: Add a blank line above this.

File: src/appleseed/foundation/utility/poison.h [hunk 2]
```diff
@@ -122,6 +135,9 @@ class PoisonImpl<double>
     {
         x = FP<double>::snan();
     }
+    static void do_always_poison(double& x){
```
Comment (oktomus) [inline src/appleseed/foundation/utility/poison.h]: Also add a blank line above this.

File: src/appleseed/foundation/utility/poison.h [hunk 3]
```diff
@@ -122,6 +135,9 @@ class PoisonImpl<double>
...
+        x = FP<double>::snan();
```
Comment (oktomus) [inline src/appleseed/foundation/utility/poison.h]: What's the difference with do_poison ?

Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: There is none, technically one could also wrap a poison call here.

File: src/appleseed/foundation/utility/poison.h (hunk 1 above)
Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: Done

File: src/appleseed/foundation/utility/poison.h (hunk 2 above)
Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: Done

File: src/appleseed/foundation/utility/poison.h [hunk 4]
```diff
@@ -82,6 +85,12 @@ class PoisonImpl
     }
 };
 
+template <typename T>
+inline void always_poison(T& x)
+{
+    PoisonImpl<T>::do_always_poison(x);
```
Comment (dictoon) [inline src/appleseed/foundation/utility/poison.h]: You should simply use `PoisonImpl<T>::do_poison(x)` here. That way we don't have to add duplicates of all the poisoning methods.

Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: I didn't do that on purpose. You never know how encapsuled the templates are. Let's say I'm using a `Vector<Vector<float, 3>>` for whatever reason (it's not the best example). In that case, the outer Vector would be called to do the `always_poison` and the inner Vector would still be stuck in the regular `poison`, resulting in no poison.
I'm relatively unhappy with this doubled infrastructure as well, but I don't want to restrict users to use built-in types at the first template-level.

**Comment by LZaw** [comment]
> I would also rename the `poison()` _free function_ to `debug_poison()` or something like that.

Will do, that's gonna take a small while, until I found all the poison calls, which have to be renamed then. :D

File: src/appleseed/foundation/utility/poison.h (hunk 4 above)
Comment (dictoon) [reply src/appleseed/foundation/utility/poison.h]: I'm not sure to understand your argument.

`poison<T>()` is just a wrapper for `PoisonImpl<T>::do_poison(x)`, except that it doesn't do anything in debug. All we want is a new variant of it that _always_ call `PoisonImpl<T>::do_poison(x)`, regardless of the configuration.

The poisoning itself is done by `PoisonImpl<T>` and it must do the same whether it's called via `debug_poison()` or `always_poison()`.

Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: Let's unwrap the chain:
In `src/appleseed/renderer/modeling/bssrdf/randomawlkbssrdf.cpp:242` the function `always_poison` gets called. This results in the following chain
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<float, 3>>::do_always_poison)
foundation/utility/poision.h:125 (PoisonImpl<float>::do_always_poison)
```
In this scenario it doesn't make a difference if we call `do_always_poison` or `do_poison`, because for floats there is no debug macro (floats will always be poisoned, regardless of the build type).
Let's assume I got a `Vector<Vector<float, 3>>` and call `poison` on it. Then the call chain would look like this:
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<Vector<float, 3>, 3>>::do_always_poison)
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:507 (PoisonImpl<Vector<float, 3>>::do_always_poison)
foundation/utility/poison.h:89 (always_poison)
foundation/utility/poision.h:125 (PoisonImpl<float>::do_always_poison)
```
This results in all floats of the inner vector to be poisoned.

But if we exchange the `do_always_poison` call for a regular `do_poison` call (as suguested), the chain looks like this:
```
foundation/utility/poison.h:89 (always_poison)
foundation/math/vector.h:500 (PoisonImpl<Vector<T, N>>::do_poison)
foundation/utility/poison.h:95 (poison)
```
The calls stop there, because the function `poison.h:95 (poison)` has the debug macro in it. We won't run into this problem, as long as the template parameter of Vectors will always be doubles or floats. But as soon as we have more complex structures, leading to a call-chain like the above mentioned one, `always_poison` will not result in a poisoned variable, which I want to prevent. From my point of view an `always` function should do its job **always**. In the above mentioned scenario it wouldn't do so.

I highly agree that this kind of double-structure is nasty and I'm thinking about how to get rid of it, without running into trouble.

The codelines of this comment refer to the as-is state in this branch. I haven't changed the `poison` functions to `debug_poison` yet. Also, I left out the `src/appleseed/` in the mutli-line structure, to increase readability.

Comment (dictoon) [reply src/appleseed/foundation/utility/poison.h]: Oh, good point, I understand now.

Maybe a nicer solution would be to ask that all `PoisonImpl` specializations use `always_poison()` in their implementation?

**Comment by LZaw** [comment]
Allright,
The old `poison` has been renamed to `debug_poison`. The `PoisonImpl<>::do_poison` now calls `always_posion`.
This way, the only difference is in the initial call to `always_poison` or `debug_poison`, with `debug_poison` having the `DEBUG` macro (and thus only doing something in debug mode).

Furthermore, I have mirrored the poison tests, so there are poison tests using `always_poison` as well. These tests will always be called, regardless of the build mode, while the `debug_poison` tests will only be available in Debug mode (just like before).
The tests have been renamed, to indicate which poison they are testing (debug or always)

All tests passed in Release build, including the "new" `always_poison` ones:
```
2019-04-17T21:44:50.965100Z <001>   113 MB info    | Foundation_Utility_Always_Poison:
2019-04-17T21:44:50.965115Z <001>   113 MB info    |   [passed] Always_Poison_Int8
2019-04-17T21:44:50.965130Z <001>   113 MB info    |   [passed] Always_Poison_UInt8
2019-04-17T21:44:50.965147Z <001>   113 MB info    |   [passed] Always_Poison_Int32
2019-04-17T21:44:50.965161Z <001>   113 MB info    |   [passed] Always_Poison_UInt32
2019-04-17T21:44:50.965173Z <001>   113 MB info    |   [passed] Always_Poison_VoidPointer
2019-04-17T21:44:50.965188Z <001>   113 MB info    |   [passed] Always_Poison_IntPointer
2019-04-17T21:44:50.965203Z <001>   113 MB info    |   [passed] Always_Poison_Float
2019-04-17T21:44:50.965220Z <001>   113 MB info    |   [passed] Always_Poison_Double
2019-04-17T21:44:50.965234Z <001>   113 MB info    |   [passed] Always_Poison_Bool
2019-04-17T21:44:50.965248Z <001>   113 MB info    |   [passed] Always_Poison_Enum

...

2019-04-17T21:45:00.041542Z <001>   123 MB info    | unit testing summary:
2019-04-17T21:45:00.041581Z <001>   123 MB info    |   suites      : 180 executed, 0 failed (0.0%)
2019-04-17T21:45:00.041581Z <001>   123 MB info    |   cases       : 1,539 executed, 0 failed (0.0%)
2019-04-17T21:45:00.041630Z <001>   123 MB info    |   assertions  : 389,581 executed, 0 failed (0.0%)
```

File: src/appleseed/foundation/utility/poison.h (hunk 4 above)
Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: Done (Details below).

File: src/appleseed/foundation/utility/poison.h [hunk 5]
```diff
@@ -65,7 +65,10 @@ namespace foundation
 //
 
 template <typename T>
-void poison(T& x);
+void debug_poison(T& x);
```
Comment (dictoon) [inline src/appleseed/foundation/utility/poison.h]: The comment above needs updating both regarding which free function to use when, and how to implement `do_poison()`.

File: src/appleseed/foundation/utility/poison.h [hunk 6]
```diff
@@ -122,6 +131,7 @@ class PoisonImpl<double>
     {
         x = FP<double>::snan();
     }
+
```
Comment (dictoon) [inline src/appleseed/foundation/utility/poison.h]: Revert this change.

File: src/appleseed/foundation/utility/poison.h (hunk 5 above)
Comment (LZaw) [reply src/appleseed/foundation/utility/poison.h]: Ah, sorry. On it.
